app/services/collections-service.js

```diff
@@ -436,3 +444,103 @@ exports.insertExport = async function(stixId, modified, exportData) {
         throw new Error(errors.notFound);
     }
 };
+
+exports.retrieveByUrl = function(url, callback) {
+    if (!url) {
+        const error = new Error(errors.missingParameter);
+        return callback(error);
+    }
+
+    superagent.get(url).then(res => {
+        try {
+            const body = JSON.parse(res.text);
+            return callback(null, body);
+        } catch (err) {
+            const error = new Error(errors.invalidFormat);
+            return callback(error);
+        }
+    }).catch(err => {
+        if (err.response && err.response.notFound) {
+            const error = new Error(errors.notFound);
+            return callback(error);
+        }
+        else if (err.response && err.response.badRequest) {
+            const error = new Error(errors.badRequest);
+            return callback(error);
+        }
+        else if (err.code === 'ENOTFOUND') {
+            const error = new Error(errors.hostNotFound);
+            return callback(error);
+        }
+        else if (err.code === 'ECONNREFUSED') {
+            const error = new Error(errors.connectionRefused);
+            return callback(error);
+        }
+        else {
+            return callback(err)
+        }
+    });
+};
+
+/**
+ * Retrieves all versions of a collection from the Workbench data store.
+ */
+function retrieveFromWorkbenchAsync(id) {
+        const options = {
+            versions: 'latest',
+        }
+        exports.retrieveById(id, options, function (err, collections) {
+            if (err) {
+                throw err;
+            }
+            else {
+                return collections;
+            }
+        });
+}
+
+/**
+ * Imports a collection bundle into the Workbench data store.
+ */
+async function importIntoWorkbenchAsync(collectionBundle, callback) {
+    try {
+        const importCollectionBundleUrl = config.workbench.restApiBaseUrl + '/api/collection-bundles';
+        const res = await authenticatedRequest
+            .post(importCollectionBundleUrl, collectionBundle);
+        const importedCollection = res.body;
+
+        return importedCollection;
+    }
+    catch(err) {
+        if (err.status === 401) {
+            throw new Error(errors.unauthorized);
+        }
+        else if (err.response && err.response.notFound) {
+            throw new Error(errors.notFound);
+        }
+        else if (err.response && err.response.badRequest) {
+            throw new Error(errors.badRequest);
+        }
+        else if (err.code === 'ENOTFOUND') {
+            throw new Error(errors.hostNotFound);
+        }
+        else if (err.code === 'ECONNREFUSED') {
+            throw new Error(errors.connectionRefused);
+        }
+        else {
+            throw err;
+        }
+    }
+}
```

Comment:
Calling the REST API endpoint from inside the REST API adds a lot of complexity to the implementation, especially in terms of making sure the configuration and authentication are correct. It is probably better to make a direct call to the service to perform the import. In this case, that would be `collectionBundlesService.importBundle()`.

There are two challenges with doing that though. The first is that the `importBundle()` call expects the collection object to have been extracted from the bundle and included as the first parameter on the call. That's not hard to do, but it does mean some extra work before calling the import function.

The second challenge is that `collectonBundlesController.importBundle()` does some important validation before calling the service, and that validation would be lost if we call the service directly. Calling the controller directly doesn't work very well--it expects the incoming data to be in an express request and it expects to finish by sending an express response. I think the best way forward will be to extract the validation logic from the controller and add it as a new function in the service. That way the controller can call it to do its validation, but the scheduler can also call it.

I'm going to go ahead and move the validation logic to the service on a new branch and merge that into `develop` when I'm done. That leaves the system in a stable state but also makes it available for you to add into this code.

Reply:
Yes, I didnt like the double REST API call either but noted all the validation logic as well, and didnt know if was out of scope.

This seems like reasonable approach though. Will wait for that merge into `develop`. Thanks.

Reply:
Refactoring the validation logic was starting to become a larger project than I had hoped. Since you're out this week and I'm out starting next week, I went ahead and made the changes necessary to call the collection bundles service directly. It doesn't include all of the validation, but it does get most of it in there. I think it's ok like that for now so I'm going to mark this ready to be merged.